### ISSUE TITLE
MapTiles download error due of CertPathValidatorException

### ISSUE THREAD

**Issue by ironspock** [streetcomplete/StreetComplete#5528]
Due previously facing crashes on SC startup I haven't updated SC to newer version than v38.2 and I'm (was still sporadically) using this version on a CAT S30 running Android 5.1 so far.

After a manual cleaning the SC cache I've download new quests. However the MapTiles weren't downloaded anymore.
Trying the same for different localtions. Allways no MapTiles download.
So for testing purposes I've updated SC version to v56.1.

No startup issues were observed with v56.1 :-)
However MapTiles weren't downloaded as well as in/with v38.2.

In the SC v56.1 logs I found following messages:
[MapTilesDownload] Error retrieving tile ...: java.security.cert.CertPathValidatorException: Trust anchor for certification path not found. 

Maybe missing certificates can fix the issue. But which ones cause the issue and how/where to get them?
Or maybe the device/OS is really too old. 

**Versions affected**
SC v56.1 (v38.2) on Android 5.1

**Comment by matkoniecz** [comment]
Yes, Android 5.1 is really old, it was released in February 2015.

**Comment by westnordost** [comment]
The issue is that some of the root certificates used today were not in the OS list of trusted root certificate authorities on early Android versions for some devices.

Maybe there are workarounds, you'll have to look for yourself.

**Comment by mnalis** [comment]
> Maybe missing certificates can fix the issue. But which ones cause the issue and how/where to get them?
Or maybe the device/OS is really too old.

Well it is old. Not certain about "too old", though, there might be chances.

**The likely issue:**

Unless I'm mistaken, the tiles are fetched from https://tile.jawg.io/ which seems to use _"Let's Encrypt"_ certificate  signed by _"ISRG Root X1"_, valid from 04 Jun 2015 to 04 Jun 2035. As such, it is probably just a little too new to have been included in your Android 5.1.

_"ISRG Root X1"_ CA was [earlier being cross-signed](https://letsencrypt.org/2020/12/21/extending-android-compatibility.html) by older _"DST Root CA X3"_ which prolonged its support on older devices, but even that extended support is [expiring now since 8 Feb 2024.](https://community.letsencrypt.org/t/shortening-the-lets-encrypt-chain-of-trust/201580).


It seems you need [at least Android 7.1.1](https://letsencrypt.org/docs/certificate-compatibility/) for automatic native support of Let's Encrypt HTTPS certificates nowdays (or will need it before the 2024 is over, if it still works for some other sites). 

See https://letsencrypt.org/docs/certificate-compatibility/ for official details.

---

**What to do, though?** 

There might be several options (from easier/better to harder) @ironspock:

- Fetching and installing _"ISRG Root X1"_ CA into your android system CA store should fix the issue. I'm unfortunately not privy to Android 5.1 device, so I don't know how (or if) you can actually do that. Perhaps just clicking on right CA `.crt` or `.der` file from your file manager might do it. Or [CAdroid app](https://f-droid.org/en/packages/at.bitfire.cadroid/) might help with it. Or you could look in thread for similar problem of Android 4.x users:  https://community.letsencrypt.org/t/help-which-certificates-to-install-on-android-4-4/212077/9 or even ask in your own thread if no other answer seems to work. 
  I'd highly **recommend** this option if at all possible.

- Alternatively, if there is no way to accept that CA as trusted by your Android; you could run your own fork of StreetComplete. It's not as hard as it looks - you need to click `fork`  on GitHub, edit [one](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/assets/map_theme/jawg/global.yaml#L144) or [two](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/screens/main/map/MapModule.kt#L24) files to specify alternative plain-http (not https) URL (i.e. `http://tile.jawg.io` seems to  be supported too on a first glance) and click on `Actions` / `Build debug APK` / `Run workflow` / choose the branch with your changes, click next `Run workflow` button, wait 15 minutes or so, and download compiled `.apk` version of StreetComplete to install on your phone. That would avoid using `https` and thus sidestep the problem. But you'd have to do that manually each time the new SC update comes out.


-  You may have heard of possibility to try to install newer (unofficial) android ROM on it, or at least root it so you can replace CAs manually yourself. However, web is full of malicious fakes, the popular valid sites like https://download.lineageos.org/devices/ don't seem to list that manufacturer/model, and even if you found some ROM there is chance you might permanently brick your phone though, and would definitely lose all your apps/data, so it's not for faint of heart. Unless you were planing to throw away that phone anyway, do not have any important data on the phone and enjoy meddling with this sort of low-level hardware-software thingies, I would skip this option. 
- Or, you can get a newer phone, obviously, but where is fun in _that_ :smiley:

**Comment by matkoniecz** [comment]
For completeness I would advise to be careful with "and install this root certificate" and "install this unofficial ROM" - if you are not understanding what is going on it is possible to end as victim of some hostile software in one or another way.

(advise by @mnalis above looks like a good one as far as I can check!)

>  However MapTiles weren't downloaded as well as in/with v38.2.

Note that it is likely to not work also there at this point - as what got broken is expiring of external certificate.

**Comment by ironspock** [comment]
As suggested by @mnalis I've installed "ISRG Root X1" CA and MapTiles download is working again.

BTW: In most cases I'm using a newer device.

Thanks for your valuable feedback!

**Comment by matkoniecz** [comment]
https://letsencrypt.org/2023/07/10/cross-sign-expiration also is relevant

**Comment by westnordost** [comment]
Google Play statistics as of 24th of March 2024:

| OS version | number of installs | in percent |
| --------------- | -------------- | --------- |
| Android 14 | 10601 | 41.0% |
| Android 13 | 6276 | 24.3% |
| Android 12 | 3172 | 12.3% |
| Android 11 | 2224 | 8.6% |
| Android 10 | 2066 | 8.0% |
| Android 9 | 752 | 2.9% |
| Android 8.1 | 132 | 0.5% |
| Android 8.0 | 209 | 0.8% |
| Android 7.1 | 58 | 0.2% |
| Android 7.0 | 100 | 0.4% |
| Android 6.0 | 37 | 0.1% |
| Android 5.1 | 25 | 0.1% |
| Android 5.0 | 13 | 0.1% |
| Android 4.4 | 2 | 0.0% |

**Comment by mnalis** [comment]
from related discussion about  Android7 and older at https://community.letsencrypt.org/t/letsencrypt-certificates-fails-on-android-phones-running-android-7-or-older/205686/2 :

> TLDR; older Android can be supported now using the "Long Chain", but you will not be able to use this chain after 2024-09-30 and must find an alternative solution before 2024-06-06.

not sure if in-app workaround for expired Let's Encrypt CA certificates from Vespucci [mentioned earlier](https://github.com/streetcomplete/StreetComplete/issues/5554#issuecomment-2163593361) at: https://github.com/MarcusWolschon/osmeditor4android/pull/2577 is also able to handle this. (someone with Android7 device might want to try if it works now, and after 2024-09-30)
